src/main/java/org/cryptomator/cryptofs/fh/ChunkCache.java

```diff
@@ -20,54 +25,166 @@ public class ChunkCache {
+	 * @throws IOException If reading or decrypting the chunk failed
+	 */
+	public Chunk getChunk(long chunkIndex) throws IOException {
+		var lock = flushLock.readLock();
+		lock.lock();
 		try {
-			return chunkLoader.load(chunkIndex);
-		} catch (AuthenticationFailedException e) {
-			// TODO provide means to pass an AuthenticationFailedException handler using an OpenOption
-			throw new IOException("Unauthentic ciphertext in chunk " + chunkIndex, e);
+			stats.addChunkCacheAccess();
+			return activeChunks.compute(chunkIndex, this::acquireInternal);
+		} catch (UncheckedIOException | AuthenticationFailedException e) {
+			throw new IOException(e);
+		} finally {
+			lock.unlock();
 		}
 	}
 
-	private void removeChunk(RemovalNotification<Long, Chunk> removal) {
+	private Chunk acquireInternal(Long index, Chunk activeChunk) throws AuthenticationFailedException, UncheckedIOException {
+		Chunk result = activeChunk;
+		if (result == null) {
+			result = staleChunks.remove(index);
+			assert result == null || result.currentAccesses().get() == 0;
+		}
+		if (result == null) {
+			result = loadChunk(index);
+		}
+
+		assert result != null;
+		result.currentAccesses().incrementAndGet();
+		return result;
+	}
+
+	private Chunk loadChunk(long chunkIndex) throws AuthenticationFailedException, UncheckedIOException {
+		stats.addChunkCacheMiss();
 		try {
-			chunkSaver.save(removal.getKey(), removal.getValue());
-			bufferPool.recycle(removal.getValue().data());
+			return new Chunk(chunkLoader.load(chunkIndex), false, () -> releaseChunk(chunkIndex));
 		} catch (IOException e) {
 			throw new UncheckedIOException(e);
 		}
 	}
 
-	public Chunk get(long chunkIndex) throws IOException {
+	@SuppressWarnings("resource")
+	private void releaseChunk(long chunkIndex) {
+		var lock = flushLock.readLock();
+		lock.lock();
 		try {
-			stats.addChunkCacheAccess();
-			return chunks.get(chunkIndex, () -> loadChunk(chunkIndex));
-		} catch (ExecutionException e) {
-			assert e.getCause() instanceof IOException; // the only checked exception thrown by #loadChunk(long)
-			throw (IOException) e.getCause();
+			activeChunks.compute(chunkIndex, (index, chunk) -> {
+				assert chunk != null;
+				var accessCnt = chunk.currentAccesses().decrementAndGet();
+				if (accessCnt == 0) {
+					staleChunks.put(index, chunk);
+					return null; //chunk is stale, remove from active
+				} else {
+					assert accessCnt > 0;
+					return chunk; //keep active
+				}
+			});
+		} finally {
+			lock.unlock();
 		}
 	}
 
-	public void set(long chunkIndex, Chunk data) {
-		chunks.put(chunkIndex, data);
+	/**
+	 * Flushes cached data (but keeps them cached).
+	 * @see #invalidateAll()
+	 */
+	public void flush() throws IOException {
+		var lock = flushLock.writeLock();
+		lock.lock();
+		BiFunction<Long, Chunk, Chunk> saveUnchecked = (index, chunk) -> {
+			try {
+				chunkSaver.save(index, chunk);
+			} catch (IOException e) {
+				throw new UncheckedIOException(e);
```

Comment:
I tested it and it turns out the mapping stays unchanged in case of an exception. Since this behaviour is not documented, I decided to add a test in 32f0ef4 to make sure we notice it, should it ever change:

https://github.com/cryptomator/cryptofs/blob/32f0ef4b93012bd0bae0ec90a5f0a2a639f60749/src/test/java/org/cryptomator/cryptofs/fh/ChunkCacheTest.java#L211-L221